[FLINK-12342][yarn] Remove container requests in order to reduce excess containers

tillrohrmann opened the pull request:
## What is the purpose of the change

This commit changes the order in which the container requests are removed when
onContainersAllocated is being called. The idea is to remove the container requests
as fast as possible in order to avoid allocating excess containers as described in
YARN-1902.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

tillrohrmann commented:
@HuangZhenQiu would it be possible for you to test this fix and verify that it solves/mitigates the problem with the excess containers significantly?

HuangZhenQiu commented:
@tillrohrmann 
Sure. We have some batch use cases that are using thousands of containers. Let me coordinate with them to test the new patch. I will ping you back once it is verified.

wangyang0918 commented:
Moving the `removeContainerRequest` first in `onContainersAllocated` is a reasonable fix. And it could reduce the excess containers to some extent. However, it could not solve the problem completely. Uploading a config file to hdfs will take 50ms in many production hdfs cluster. It means only 10 containers could be started. If we receive more than 10 containers, the `CallbackHandlerThread` of yarn am-rm client will be blocked. So `removeContainerRequest` may not be called timely.

wangyang0918 commented:
I have tested the PR in the following cases.
Ask 1000 containers with 500ms `containerRequestHeartbeatInterval`
* All containers are returned in one heart beat. There is no excess containers.
* Containers are returned in two or more heart beats. More than 3000 containers are allocated from yarn. It is based whether the hdfs is busy.

Anyway, this PR is useful especially when the yarn cluster has enough resource.
I am in favor of merging.

tillrohrmann commented:
> Moving the `removeContainerRequest` first in `onContainersAllocated` is a reasonable fix. And it could reduce the excess containers to some extent. However, it could not solve the problem completely. Uploading a config file to hdfs will take 50ms in many production hdfs cluster. It means only 10 containers could be started. If we receive more than 10 containers, the `CallbackHandlerThread` of yarn am-rm client will be blocked. So `removeContainerRequest` may not be called timely.

I think we don't execute the removal of the container requests and the upload of files in the `CallbackHandlerThread` but in the `YarnResourceManager's` main thread. Hence, we should call `removeContainerRequest` fast enough once we start processing the callback within the main thread.

>* All containers are returned in one heart beat. There is no excess containers.
>* Containers are returned in two or more heart beats. More than 3000 containers are allocated from yarn. It is based whether the hdfs is busy.

Could you explain the second point a bit more. How did you provoke that the containers are returned in two or more heartbeats? Was the `ResourceManager` busy with other things?

tillrohrmann commented:
There was a small bug in the PR @wangyang0918. Please take another look. It might have affected your test run.

zentol commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -370,45 +370,24 @@ public void onContainersCompleted(final List<ContainerStatus> statuses) {
-					"Received new container: {} - Remaining pending container requests: {}",
-					container.getId(),
-					numPendingContainerRequests);
+			final int numAcceptedContainers = Math.min(containers.size(), numPendingContainerRequests);
```
maybe add a comment that we use the minimum since the number of containers can exceed the number of pending requests.

tillrohrmann replied:
Will do.

zentol commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -370,45 +370,24 @@ public void onContainersCompleted(final List<ContainerStatus> statuses) {
-					} catch (Throwable t) {
-						log.error("Could not start TaskManager in container {}.", container.getId(), t);
+			for (Container excessContainer : excessContainers) {
+				returnExcessContainer(excessContainer);
```
`excessContainer.forEach(this::returnExcessContainer)`?

zentol commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -370,45 +370,24 @@ public void onContainersCompleted(final List<ContainerStatus> statuses) {
-							containerIdStr,
-							container.getNodeId().getHost());
+			for (int i = 0; i < requiredContainers.size(); i++) {
+				removeContainerRequest(pendingRequestsIterator.next());
```
`requiredContainers.forEach(c -> removeContainerRequest(pendingRequestsIterator.next()));`?

tillrohrmann replied:
I think the for loop produces fewer objects and is not less expressive.

zentol commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -419,6 +398,36 @@ public void onContainersAllocated(List<Container> containers) {
 		});
 	}
 
+	private void startTaskExecutorInContainer(Container requiredContainer) {
```
wondering whether the `required` prefix makes sense in the context of this method; I'd tend to say no.

tillrohrmann replied:
True, will change it.

zentol commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -370,45 +370,24 @@ public void onContainersCompleted(final List<ContainerStatus> statuses) {
 	@Override
 	public void onContainersAllocated(List<Container> containers) {
 		runAsync(() -> {
```
Is there possibly any value in moving part of this method out of the async block? (removing requests and returning containers)

The idea being that under heavy loadcthe processing of the allocation could be delayed for quite a while.

tillrohrmann replied:
For this, we need to read `numPendingContainerRequests` which can only be done from within the actor's main thread. Hence without introducing some `volatiles` and potential race conditions I don't see an easy way.

zentol commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -370,45 +370,24 @@ public void onContainersCompleted(final List<ContainerStatus> statuses) {
-					resourceManagerClient.releaseAssignedContainer(container.getId());
-				}
+			for (Container requiredContainer : requiredContainers) {
+				startTaskExecutorInContainer(requiredContainer);
```
Do we have any info on how often this call fails? (we could reuse an excess container if this fails, which would save us from having to re-allocate another one)

tillrohrmann replied:
I don't think that this happens too often. If this should become a problem, then I would suggest to apply this optimization. At the moment it would only complicate the implementation.

tillrohrmann commented:
Thanks for your review @zentol. I've addressed your comments and rebased onto the latest master. Concerning using some of the excess containers in case of a `TaskExecutor` start failure, I would suggest to leave this as a future optimization since it complicates the implementation adding a higher maintenance burden.

I've rebased the PR onto the latest master and squashed the fixups.

wangyang0918 commented:
@tillrohrmann 
1. You are right. The the removal of the container requests and the upload of files is running in the `YarnResourceManager`'s main thread. I think it is a single thread, so if we receive 100 containers in the first heart beat, and it will take 5000ms to launch them(50ms for each). Even we receive all the remaining 900 containers in the second heart beat, `removeContainerRequest` could not be called timely. So we still have excess containers. I agree with you that it could be done in the future optimization. (1) Do not upload taskmanager-conf.yaml to hdfs, use dynamic properties instead. (2) Use `NMClientAsync` instead of `NMClient`.
2. I start some other jobs to occupy most resources of the yarn cluster. When i have submitted the flink cluster, i kill some of them one by one. So the required 1000 containers are allocated in two or more heart beats.

tillrohrmann commented:
> 1. You are right. The the removal of the container requests and the upload of files is running in the `YarnResourceManager`'s main thread. I think it is a single thread, so if we receive 100 containers in the first heart beat, and it will take 5000ms to launch them(50ms for each). Even we receive all the remaining 900 containers in the second heart beat, `removeContainerRequest` could not be called timely. So we still have excess containers. I agree with you that it could be done in the future optimization. (1) Do not upload taskmanager-conf.yaml to hdfs, use dynamic properties instead. (2) Use `NMClientAsync` instead of `NMClient`.

All right, now understand the problem @wangyang0918. Yes if the main thread is blocked (this applies to other blocking RM operations as well), then it might happen that we don't process another `onContainersAllocated` in time. I think the general solution is to avoid blocking operations being executed in the main thread. I would suggest to change https://issues.apache.org/jira/browse/FLINK-14582 in this direction and let avoiding to write a file being one solution approach. As you've said there are other approaches as well, like using a dedicated thread or maybe the `NMClientAsync`. In fact the last approach has already been proposed in https://issues.apache.org/jira/browse/FLINK-13184.

roshannaik commented:
Was directed to this patch by @HuangZhenQiu.  Applied this to Flink 1.6 and used this to launch a job with 500 yarn containers. See definite improvement with the overallocation problem. However once the job comes up (as yet yarn ui), the Flink UI fails to load. 

Perhaps the UI is having trouble with large number of containers ?

tillrohrmann commented:
Hard to tell why the UI hangs. I'm not aware of specific improvements apart from the fact that it has been completely reworked using a different JavaScript framework. You could take a look at the logs to see whether something suspicious is visible.

In order to fully solve the problem of too many excess containers, please also take a look at FLINK-13184 and FLINK-15036.

roshannaik commented:
What logs are you referring to ? Are there any UI related logs in AppMaster ?

Thanks for the pointers to those two tickets.

tillrohrmann commented:
I'm referring to the cluster entry point logs. They should be either called `jobmanager.log` or `standalonesession.log`.
